pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/IdenticalCatchBranchesRule.java

```diff
@@ -0,0 +1,174 @@
+        return tokenEquals(st1.getBlock(), st2.getBlock(), st1.getExceptionName(), st2.getExceptionName());
+    }
+
+
+    // groups catch statements by equivalence class, according to the string value of their tokens.
+    private Set<List<ASTCatchStatement>> equivalenceClasses(List<ASTCatchStatement> catches) {
+        Set<List<ASTCatchStatement>> result = new HashSet<>(catches.size());
+        for (ASTCatchStatement stmt : catches) {
+            if (result.isEmpty()) {
+                result.add(newEquivClass(stmt));
+                continue;
+            }
+
+            boolean isNewClass = true;
+            for (List<ASTCatchStatement> equivClass : result) {
+                if (areEquivalent(stmt, equivClass.get(0))) {
+                    equivClass.add(stmt);
+                    isNewClass = false;
+                    break;
+                }
+            }
+
+            if (isNewClass) {
+                result.add(newEquivClass(stmt));
+            }
+        }
+
+        return result;
+    }
+
+
+    private List<ASTCatchStatement> newEquivClass(ASTCatchStatement stmt) {
+        // Each equivalence class is sorted by document order
+        List<ASTCatchStatement> result = new ArrayList<>(2);
+        result.add(stmt);
+        return result;
+    }
+
+
+    // Gets the representation of the set of catch statements as a single multicatch
+    private String getCaughtExceptionsAsString(ASTCatchStatement stmt) {
+
+        StringBuilder sb = new StringBuilder();
+
+        final String delim = " | ";
+        for (ASTType type : stmt.getCaughtExceptionTypeNodes()) {
+            sb.append(type.getTypeImage()).append(delim);
+        }
+
+        // remove the last delimiter
+        sb.replace(sb.length() - 3, sb.length(), "");
+        return sb.toString();
+    }
+
+
+    @Override
+    public Object visit(ASTTryStatement node, Object data) {
+
+        List<ASTCatchStatement> catchStatements = node.getCatchStatements();
+        Set<List<ASTCatchStatement>> equivClasses = equivalenceClasses(catchStatements);
+
+        for (List<ASTCatchStatement> identicalStmts : equivClasses) {
+            if (identicalStmts.size() > 1) {
+                String identicalBranchName = getCaughtExceptionsAsString(identicalStmts.get(0));
+
+                // By convention, lower catch blocks are collapsed into the highest one
+                // The first node of the equivalence class is thus the block that should be transformed
+                for (int i = 1; i < identicalStmts.size(); i++) {
+                    addViolation(data, identicalStmts.get(i), new String[]{identicalBranchName, });
+                }
+            }
+        }
+
+        return super.visit(node, data);
+    }
+
+
+    // Would be cleaner with java8 streams. Maybe this can be moved to AbstractJavaNode then
+    private static Iterator<String> tokenImageIterator(final AbstractJavaNode node) {
```

Comment:
@oowekyala If we want to implement subtree comparison, I think `tokenImageIterator` and `tokenEquals` should be refactored and logic will be too complicate. So I just make a little changes in `tokenEquals`. If I misunderstand it, please correct me:)

Reply:
You could just delete tokenEquals and tokenImageIterator and plug your original subtree comparison routine into `areEquivalent`. Token image comparison was just a shortcut to determine the equivalence of two subtrees, but you did implement the real thing ;)

Reply:
Done.